### PR TITLE
Feature - Refactoring @Field decorator

### PR DESCRIPTION
**Pull request by felipesabino**
Sorry but this will be a big PR with a lot of breaking changes

---

- [x] Refactored `@Field` decorator and nested dependencies (`@Arg`, `@OrderBy`, `@Root`, `@Before` and `@Ctx`)  - #22 
- [x] Renamed `DefaultOption`, `DefaultMetadata` and `DefaultArg` interfaces to just `Option`, `Metadata` and `Argument` instead as in https://github.com/indigotech/graphql-schema-decorator/pull/32#discussion_r147517577
- [x] Refactored metadata handling (to use `MetadataBuilder` instead of `Reflect`)
- [x] Refactored `field_type_factory` to handle metadata properly and to be less verbose/confusing
- [x] Added functional tests (to make sure resolvers are working properly)
- [x] Removed old (and commented) code
- [x] Updated examples
- [x] Review JSDoc for each decorator and its option argument - #23 
- [x] Review `TODO`s for possible missing tests
- [x] Deprecated all usages of 
  - `@Pagination`
  - `@Description`
  - `@List`
  - `@NonNull`

### CODE REVIEW

File: src/pagination.middleware.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import 'reflect-metadata';
```
Comment (tibawatanabe) [inline src/pagination.middleware.ts]: Do we need this here? I'm not finding a need...

**Review by tibawatanabe — APPROVED** [review]
2 comments:

- I think there are some unneeded imports of `reflect-metadata` 
- `ArgumentArg`... a little bit strange but I have no better suggestion

File: src/pagination.middleware.ts (hunk 1 above)
Comment (felipesabino) [reply src/pagination.middleware.ts]: > I think there are some unneeded imports of reflect-metadata

I was getting a lot of error like

```
Uncaught TypeError: Reflect.getMetadata is not a function
```

So this helped me suppress it... found this solution [here](https://github.com/AngularClass/angular-starter/issues/1238#issuecomment-268723687) but do not know a better alternative.

Do you know the root cause for it? Any idea if something at `tsconfig` might fix this issue without these imports?

> `ArgumentArg`... a little bit strange but I have no better suggestion

Same here 😞

**Review by thiago-soliveira — APPROVED** [review]

File: src/pagination.middleware.ts (hunk 1 above)
Comment (felipesabino) [reply src/pagination.middleware.ts]: @tibawatanabe 

> I think there are some unneeded imports of reflect-metadata

Ok, I did some reading and I need it just in my app/lib entry point, so I added it to both `index.ts` and `*.spec.ts`

**Pull request merged**